server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java

```diff
@@ -8,54 +8,112 @@
 
 package org.elasticsearch.search.profile.query;
 
+import org.apache.lucene.sandbox.search.ProfilerCollector;
 import org.apache.lucene.search.Collector;
 import org.apache.lucene.search.CollectorManager;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.List;
-import java.util.stream.Collectors;
 
 /**
  * A {@link CollectorManager} that takes another CollectorManager as input and wraps all Collectors generated by it
  * in an {@link InternalProfileCollector}. It delegates all the profiling to the generated collectors via {@link #getCollectorTree()}
- * and joins them up when its {@link #reduce} method is called. The profile result can
+ * and joins the different collector trees together when its {@link #reduce} method is called.
+ * Supports optionally providing sub-collector managers for top docs as well as aggs collection, so that each
+ * {@link InternalProfileCollector} created is provided with the corresponding sub-collectors that are children of the top-level collector.
+ * @param <T> the return type of the wrapped collector manager, which the reduce method returns.
  */
 public final class ProfileCollectorManager<T> implements CollectorManager<InternalProfileCollector, T> {
 
-    private final CollectorManager<Collector, T> collectorManager;
+    private final CollectorManager<? extends Collector, T> collectorManager;
     private final String reason;
+    private final ProfileCollectorManager<?> topDocsSubCollectorManager;
+    private final ProfileCollectorManager<?> aggsSubCollectorManager;
+    // this is a bit of a hack: it allows us to retrieve the last collector that newCollector has returned for sub-collector managers,
+    // so that we can provide them to InternalProfileCollector's constructor as children.
+    private InternalProfileCollector profileCollector;
```

Comment:
If I'm not mistaken this is problematic in a concurrent setting. If multiple threads call this managers newCollector(), assuming this is a MultiCollectorManager, that will propagate that to its children. How can we be sure another thread doesn't do this simultaneously and the child "profileCollector" is already updated to the collector created by that other thread?

Reply:
it is not because newCollector is called before parallelizing. I had a comment on this before, maybe I should add it back for clarify.

Reply:
I see. That requires knowledge of the calling code though and introduces a dependency on that usage strategy. Not ideal but workable atm.

Reply:
It relies on the contract of CollectorManagers: the collector manager provides the collectors, the collection is parallelized, then reduce is called providing those same collectors. It does require knowledge of how it's used but it is not part of the contract that e.g. each thread calls newCollector to retrieve its collector, because then it would not be possible to call reduce providing all the collectors from all threads.